File: resource-server/app/services/topology.py
from app.models.topology import Topology


def format_topology_markdown(topology: Topology) -> str:
    if not topology.nodes:
        return ""

    lines = ["## Service Topology", ""]

    adjacency: dict[str, list[str]] = {}
    for edge in topology.edges:
        adjacency.setdefault(edge.from_, []).append(edge.to)

    for node in topology.nodes:
        label = node.id
        if node.engine:
            label += f" ({node.engine})"
        elif node.kind != "service":
            label += f" ({node.kind})"
        deps = adjacency.get(node.id) or []
        if deps:
            dep_labels = []
            for dep_id in deps:
                dep_node = next((n for n in topology.nodes if n.id == dep_id), None)
                if dep_node and dep_node.engine:
                    dep_labels.append(f"{dep_id} ({dep_node.engine})")
                else:
                    dep_labels.append(dep_id)
            lines.append(f"- {label} → {', '.join(dep_labels)}")
        else:
            lines.append(f"- {label}")

    if topology.targets:
        lines.append("")
        target_bits = [f"{target.id}={target.type}" for target in topology.targets]
        lines.append(f"- targets: {', '.join(target_bits)}")

    return "\n".join(lines)


def deployment_summary(topology: Topology | None) -> dict:
    if topology is None:
        return {
            "platform": "unknown",
            "services": [],
            "nodeCount": 0,
            "targets": [],
        }

    platform = "unknown"
    if topology.targets:
        primary = topology.targets[0].type
        if primary == "docker-compose":
            platform = "docker-compose"
        elif primary == "kubernetes":
            platform = "kubernetes"
        else:
            platform = "unknown"
    elif any(node.kind == "service" for node in topology.nodes):
        platform = "docker-compose"

    services = [
        {
            "name": node.id,
            "kind": node.kind,
            "image": node.image,
            "ports": [
                f"{p.host}:{p.container}" if p.host is not None else str(p.container)
                for p in node.ports
            ],
        }
        for node in topology.nodes
        if node.kind in ("service", "gateway", "worker")
    ]

    return {
        "platform": platform,
        "services": services,
        "nodeCount": len(topology.nodes),
        "targets": [target.model_dump() for target in topology.targets],
    }

File: desktop/electron/resources/topology.ts
import fs from "node:fs/promises";
import path from "node:path";
import yaml from "yaml";

export interface TopologyPort {
  container: number;
  host?: number;
}

export interface TopologyNode {
  id: string;
  kind: "service" | "database" | "cache" | "gateway" | "worker";
  runtime?: string;
  engine?: string;
  image?: string | null;
  source?: string;
  dockerfile?: string;
  ports?: TopologyPort[];
}

export interface TopologyEdge {
  from: string;
  to: string;
  env?: Record<string, string>;
}

export interface TopologyTarget {
  id: string;
  type: "docker-compose" | "kubernetes" | "paas";
  env?: string;
}

export interface Topology {
  version: 1;
  project: string;
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  targets: TopologyTarget[];
}

export async function fetchServerTopology(
  serverUrl: string,
  projectId?: string,
): Promise<Topology | null> {
  try {
    const base = serverUrl.replace(/\/$/, "");
    const query = projectId ? `?project=${encodeURIComponent(projectId)}` : "";
    const res = await fetch(`${base}/v1/topology${query}`);
    if (res.status === 404) {
      return null;
    }
    if (!res.ok) {
      return null;
    }
    return (await res.json()) as Topology;
  } catch {
    return null;
  }
}

export async function loadLocalTopology(projectRoot: string): Promise<Topology | null> {
  const topologyPath = path.join(projectRoot, ".agentflow/topology.yaml");
  try {
    const raw = await fs.readFile(topologyPath, "utf8");
    return yaml.parse(raw) as Topology;
  } catch {
    return null;
  }
}

export function projectIdFromRoot(projectRoot: string): string {
  return path.basename(projectRoot);
}

export async function resolveTopology(
  projectRoot: string,
  resourceServerUrl?: string,
  projectId?: string,
): Promise<Topology | null> {
  const local = await loadLocalTopology(projectRoot);
  if (local) {
    return local;
  }
  if (!resourceServerUrl) {
    return null;
  }
  const id = projectId ?? projectIdFromRoot(projectRoot);
  return fetchServerTopology(resourceServerUrl, id);
}

export function formatTopologyContextForPrompt(topology: Topology): string {
  if (!topology.nodes.length) {
    return "";
  }

  const lines = ["## Service Topology", ""];
  const adjacency = new Map<string, string[]>();

  for (const edge of topology.edges) {
    const list = adjacency.get(edge.from) ?? [];
    list.push(edge.to);
    adjacency.set(edge.from, list);
  }

  for (const node of topology.nodes) {
    let label = node.id;
    if (node.engine) {
      label += ` (${node.engine})`;
    } else if (node.kind !== "service") {
      label += ` (${node.kind})`;
    }
    if (node.source) {
      label += ` [source: ${node.source}]`;
    }
    const deps = adjacency.get(node.id);
    if (deps?.length) {
      const depLabels = deps.map((depId) => {
        const depNode = topology.nodes.find((n) => n.id === depId);
        if (depNode?.engine) {
          return `${depId} (${depNode.engine})`;
        }
        return depId;
      });
      lines.push(`- ${label} → ${depLabels.join(", ")}`);
    } else {
      lines.push(`- ${label}`);
    }
  }

  if (topology.targets.length) {
    lines.push("");
    lines.push(
      `- targets: ${topology.targets.map((t) => `${t.id}=${t.type}`).join(", ")}`,
    );
  }

  return lines.join("\n");
}

export function combineResourceAndTopologyContext(
  resourceMarkdown: string,
  topologyMarkdown: string,
): string {
  return [resourceMarkdown, topologyMarkdown].filter((part) => part.trim()).join("\n\n");
}
